File: tests/demo-transport.test.ts
import {afterEach, beforeEach, describe, expect, test, vi} from "vitest";

import {DemoTransport} from "../src/demo-transport.ts";
import type {ZulipEvent} from "../src/types.ts";

describe("DemoTransport", () => {
    beforeEach(() => {
        vi.useFakeTimers();
    });

    afterEach(() => {
        vi.useRealTimers();
    });

    test("seeds initial messages and reports connected", async () => {
        const transport = new DemoTransport({
            scope: {channel: "general", topic: "hello"},
            autoReply: false,
        });
        const events: ZulipEvent[] = [];
        await transport.connect((e) => events.push(e));
        const messages = await transport.getMessages({channel: "general", topic: "hello"});

        expect(messages.length).toBeGreaterThan(0);
        expect(events.at(0)).toEqual({type: "connection", status: "connected"});
        await transport.close();
    });

    test("sendMessage appends and emits a message event", async () => {
        const transport = new DemoTransport({
            scope: {channel: "general", topic: "hello"},
            autoReply: false,
        });
        const received: ZulipEvent[] = [];
        await transport.connect((e) => received.push(e));

        await transport.sendMessage({
            type: "channel",
            channel: "general",
            topic: "hello",
            content: "Hi team",
        });

        const messageEvents = received.filter((e) => e.type === "message");
        expect(messageEvents).toHaveLength(1);
        expect(messageEvents[0]).toMatchObject({
            type: "message",
            message: {content: "Hi team"},
        });
        await transport.close();
    });

    test("autoReply emits a second message after the delay", async () => {
        const transport = new DemoTransport({
            scope: {channel: "general", topic: "hello"},
            autoReply: true,
            autoReplyDelayMs: 500,
        });
        const received: ZulipEvent[] = [];
        await transport.connect((e) => received.push(e));

        await transport.sendMessage({
            type: "channel",
            channel: "general",
            topic: "hello",
            content: "Ping?",
        });
        vi.advanceTimersByTime(500);

        const messageEvents = received.filter((e) => e.type === "message");
        expect(messageEvents).toHaveLength(2);
        expect(messageEvents[1]?.message.senderEmail).toBe("zulip-bot@example.com");
        await transport.close();
    });

    test("close cancels pending auto-replies", async () => {
        const transport = new DemoTransport({
            scope: {channel: "general"},
            autoReply: true,
            autoReplyDelayMs: 500,
        });
        const received: ZulipEvent[] = [];
        await transport.connect((e) => received.push(e));
        await transport.sendMessage({type: "channel", channel: "general", content: "Hello?"});

        await transport.close();
        vi.advanceTimersByTime(1_000);

        const messageEvents = received.filter((e) => e.type === "message");
        expect(messageEvents).toHaveLength(1);
    });
});

File: packages/flutter/test/demo_transport_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:zulip_embed/zulip_embed.dart';

void main() {
  group('DemoTransport', () {
    test('reports connecting then connected on connect()', () async {
      final transport = DemoTransport();
      final events = <ZulipEvent>[];
      await transport.connect(
        scope: const ScopeFilter(channel: 'general', topic: 'welcome'),
        onEvent: events.add,
      );
      final statuses =
          events.whereType<ConnectionEvent>().map((e) => e.status).toList();
      expect(statuses.first, ConnectionStatus.connecting);
      expect(statuses.last, ConnectionStatus.connected);
      await transport.close();
    });

    test('seeds welcome messages scoped to the channel/topic', () async {
      final transport = DemoTransport();
      await transport.connect(
        scope: const ScopeFilter(channel: 'design', topic: 'intro'),
        onEvent: (_) {},
      );
      final messages = await transport.getMessages(
        scope: const ScopeFilter(channel: 'design', topic: 'intro'),
      );
      expect(messages.length, greaterThanOrEqualTo(2));
      expect(
        messages.every(
          (m) => m is ChannelMessage && m.channelName == 'design',
        ),
        isTrue,
      );
      expect(
        messages.every((m) => m is ChannelMessage && m.topic == 'intro'),
        isTrue,
      );
      await transport.close();
    });

    test('sendMessage emits the sent message and an echo reply', () async {
      final transport = DemoTransport();
      final events = <ZulipEvent>[];
      await transport.connect(
        scope: const ScopeFilter(channel: 'general', topic: 'welcome'),
        onEvent: events.add,
      );
      await transport.sendMessage(
        const ChannelSendParams(
          channel: 'general',
          topic: 'welcome',
          content: 'Hello Zulip',
        ),
      );
      await Future<void>.delayed(const Duration(milliseconds: 1000));
      final messages =
          events.whereType<MessageEvent>().map((e) => e.message).toList();
      expect(messages.length, greaterThanOrEqualTo(2));
      expect(messages.first.content, 'Hello Zulip');
      expect(messages[1].content, contains('Echo: Hello Zulip'));
      await transport.close();
    });

    test('close() cancels any pending echo replies', () async {
      final transport = DemoTransport();
      final events = <ZulipEvent>[];
      await transport.connect(
        scope: const ScopeFilter(channel: 'general'),
        onEvent: events.add,
      );
      await transport.sendMessage(
        const ChannelSendParams(
          channel: 'general',
          topic: 'general chat',
          content: 'hi',
        ),
      );
      final before = events.whereType<MessageEvent>().length;
      await transport.close();
      await Future<void>.delayed(const Duration(milliseconds: 1000));
      final after = events.whereType<MessageEvent>().length;
      expect(after, before);
    });
  });

  group('ZulipClient', () {
    test('rebroadcasts transport events via a broadcast stream', () async {
      final transport = DemoTransport();
      final client = ZulipClient(transport);
      final received = <ZulipEvent>[];
      final sub = client.events.listen(received.add);
      await client.connect(const ScopeFilter(channel: 'general'));
      await client.sendMessage('ping');
      await Future<void>.delayed(const Duration(milliseconds: 1000));
      await sub.cancel();
      await client.dispose();
      expect(received.whereType<ConnectionEvent>(), isNotEmpty);
      expect(received.whereType<MessageEvent>().length, greaterThanOrEqualTo(2));
    });
  });
}
